pkg/distributor/forwarding/forwarding.go

```diff
@@ -290,20 +343,30 @@ func (r *request) sendToEndpoint(ctx context.Context, endpoint string, ts []mimi
 		r.errors.WithLabelValues(strconv.Itoa(httpResp.StatusCode)).Inc()
 		err := errors.Errorf("server returned HTTP status %s: %s", httpResp.Status, line)
 		if httpResp.StatusCode/100 == 5 || httpResp.StatusCode == http.StatusTooManyRequests {
-			return recoverableError{err}
+			// The forwarding endpoint has returned a retriable error, so we want the client to retry.
+			r.handleError(http.StatusInternalServerError, err)
+			return
 		}
-		return err
+		r.handleError(http.StatusBadRequest, err)
 	}
+}
 
-	return nil
+func (r *request) handleError(status int, err error) {
+	errMsg := err.Error()
+	level.Warn(r.log).Log("msg", "error in forwarding request", "err", errMsg)
+	if r.propagateErrors {
+		r.errCh <- httpgrpc.Errorf(status, errMsg)
+	}
 }
 
-// cleanup must be called to return the used buffers to their pools after a request has completed.
 func (r *request) cleanup() {
-	for _, ts := range r.tsByEndpoint {
-		*ts = (*ts)[:0]
-		r.pools.timeseries.Put(ts)
-	}
+	r.pools.putTsSlice(r.ts.ts)
+
+	// Ensure we don't modify a request property after returning it to the pool by calling .Done() on the wait group.
+	wg := r.requestWg
+	r.requestWg = nil
+
+	r.pools.putReq(r)
 
-	r.tsByEndpoint = nil
+	wg.Done()
```

Comment:
Why can't we just do this?
```go
r.requestWg.Done()
r.pools.putReq(r)
```

Reply:
Simply for testability :)

If the object would be returned to the pool after the request wait group is done then there is no way to ensure that we only validate the pool usage after all objects have been returned to the pool. If you have a better idea how to handle that then please let me know.

For now I'll add a comment mentioning this.

Reply:
added this comment: 59784456a45e2e190689b9f2e03e6b491ad8ab45